### ISSUE TITLE
damage_all_outputs / wlr-scene

### ISSUE THREAD

**Issue by Consolatis** [labwc/labwc#230]
During the last days I was on a quest to eliminate nearly every single `damage_all_outputs()` call but its actually tougher than expected.
The best I came up with is replacing the view related calls with `damage_view_outputs()` which still damage whole outputs but at least only those outputs where (parts of) the view is displayed on. Menu isn't that easy either because it can span over multiple outputs and we don't really track its full dimensions. I still don't really like any of the approaches I took to only damage what really needs to be marked as damaged (resizing / moving is a mess).
Another approach I didn't really investigate yet is something like `damage_layout_box(struct wlr_box *box)` which would automatically damage the appropiate output boxes but its actually pretty messy to get the full extends of a view every time we want to damage a view.

So I had a quick look at `wlr_scene`, more specifically at `wlroots/examples/scene-graph.c` and changed it to something like this:
```
#define SSD_HEIGHT 26
#define BORDER_WIDTH 3
struct surface {
	struct wlr_surface *wlr_surface;
	struct scene_tree *scene_toplevel;
	struct scene_rect *scene_ssd;
	struct scene_rect *scene_border;
	struct scene_node *scene_surfaces;
	..
};
```

This is the scene layout:
```
scene
  +--scene_toplevel_1
       +--scene_ssd (0, 0, wlr_surface->width + 2 * BORDER_WIDTH, SSD_HEIGHT)
       +--scene_border (0, SSD_HEIGHT, wlr_surface->width + 2 * BORDER_WIDTH, wlr_surface->height + BORDER_WIDTH)
       +--scene_surfaces (BORDER_WIDTH, SSD_HEIGHT, wlr_surface->width, wlr_surface->height)
            +--[subsurfaces]
  +--scene_toplevel_2
  ..
```

`scene_toplevel` can then be used with `wlr_scene_node_set_position(surface->scene_toplevel->node, x, y)` to move the whole toplevel including its test SSD `scene_rect`.
On `wlr_surface` dimension change, `scene_ssd` and `scene_border` sizes are updated to match the new offsets.

I didn't investigate at all how mouse input works with `wlr_scene`.
And there doesn't seem to be an easy way to get the full recursive dimensions of a given tree node either..

So far I am not sure if wlr-scene actually provides substantial benefits but maybe I am missing something.
Is there a public branch somewhere with early dev work for scene-graph?

**Comment by droc12345** [comment]
Did you look at tinywl in the wlroots git.

https://gitlab.freedesktop.org/wlroots/wlroots/-/tree/master/tinywl

**Comment by Consolatis** [comment]
Thanks for the links.
I did look into tinywl but that must have been before it was changed to using the scene-graph API so I was under the impression tinywl wasn't using it. Tinywl indeed provides a much better testground to play around with scene-graph.

The issue we have here though is that there are some functions in labwc (like SSD and menu) which are rendered outside of wlroots that have to be damaged as well. The menu could likely be converted into a scene node which would indeed make it easier to close the menu, we'd just disable the menu root node and the internal damage tracking of scene-graph should take care of the rest. For hover effects to open a submenu and changing the background color though I am not sure if scene-graph would help. Same for server side decorations.

So far I think there is something missing from scene-graph: an explicit `scene_damage_node(bool recursive)` or something along those lines.
